### ISSUE TITLE
Task-augmented tool call errors are wrapped incorrectly, masking actual error messages

### ISSUE THREAD

**Issue by punkpeye** [modelcontextprotocol/typescript-sdk#1385]
### Summary

When a task-augmented `tools/call` request fails (e.g., due to input validation), the error is caught and converted to a tool error result (`{content: [...], isError: true}`), which then fails `CreateTaskResultSchema` validation. This results in a confusing error message that masks the actual underlying error.

### Steps to Reproduce

1. Register a task-based tool with input validation:

```typescript
server.experimental.tasks.registerToolTask(
  'batch_process',
  {
    inputSchema: {
      itemCount: z.number().min(1).max(10),
      processingTimeMs: z.number().min(500).max(5000).optional(),
    },
    // ... other config
  },
  {
    createTask: async (args, extra) => {
      // ... handler
    },
  },
);
```

2. Call the tool with task augmentation and invalid arguments:

```json
{
  "method": "tools/call",
  "params": {
    "name": "batch_process",
    "arguments": { "itemCount": 5, "processingTimeMs": 100 },
    "task": { "ttl": 60000 }
  }
}
```

### Expected Behavior

The client should receive a clear error message indicating the actual problem:

```json
{
  "error": {
    "code": -32602,
    "message": "Input validation error: Invalid arguments for tool batch_process: Too small: expected number to be >=500"
  }
}
```

### Actual Behavior

The client receives a confusing error that hides the actual cause:

```json
{
  "error": {
    "code": -32602,
    "message": "MCP error -32602: Invalid task creation result: [{\"expected\":\"object\",\"code\":\"invalid_type\",\"path\":[\"task\"],\"message\":\"Invalid input: expected object, received undefined\"}]"
  }
}
```

### Root Cause

The issue is in the error handling flow in `server/mcp.js`:

```javascript
// In setToolRequestHandlers() - CallToolRequestSchema handler
try {
  const args = await this.validateToolInput(tool, request.params.arguments, request.params.name);
  const result = await this.executeToolHandler(tool, args, extra);
  // ...
} catch (error) {
  // ALL errors get wrapped as tool errors, including validation errors
  return this.createToolError(error instanceof Error ? error.message : String(error));
}
```

Then in `server/index.js`, the wrapped handler validates the result:

```javascript
if (params.task) {
  const taskValidationResult = safeParse(CreateTaskResultSchema, result);
  if (!taskValidationResult.success) {
    // This fails because createToolError returns {content: [...], isError: true}
    // which doesn't have a 'task' property
    throw new McpError(ErrorCode.InvalidParams, `Invalid task creation result: ${errorMessage}`);
  }
}
```

### Suggested Fix

For task-augmented requests, errors thrown before task creation (like input validation errors) should be re-thrown directly instead of being wrapped as tool errors:

```javascript
// In setToolRequestHandlers() - CallToolRequestSchema handler
try {
  // ...
} catch (error) {
  if (error instanceof McpError) {
    if (error.code === ErrorCode.UrlElicitationRequired) {
      throw error;
    }
    // For task-augmented requests, also re-throw validation errors
    // instead of wrapping them as tool errors
    if (isTaskRequest && error.code === ErrorCode.InvalidParams) {
      throw error;
    }
  }
  return this.createToolError(error instanceof Error ? error.message : String(error));
}
```

Or alternatively, modify `server/index.js` to not validate non-task results against `CreateTaskResultSchema`:

```javascript
if (params.task) {
  // Only validate if the result looks like a CreateTaskResult (has task property)
  // If it's a tool error, return it as-is or convert to a proper error response
  if ('isError' in result && result.isError) {
    throw new McpError(ErrorCode.InvalidParams, result.content[0]?.text ?? 'Tool execution failed');
  }
  // ... existing validation
}
```

### Environment

- SDK Version: 1.25.2
- Node.js: v24.1.0

---

This bug affects the developer experience significantly as the actual error (input validation, missing taskStore, handler exceptions, etc.) is completely hidden behind the generic "Invalid task creation result" message.

**Comment by punkpeye** [comment]
you can replicate this in an inspector

https://glama.ai/mcp/inspector?servers=%5B%7B%22id%22%3A%22test%22%2C%22name%22%3A%22test%22%2C%22requestTimeout%22%3A10000%2C%22url%22%3A%22https%3A%2F%2Fmcp-test.glama.ai%2Fmcp%22%7D%5D

select batch_process and then enter invalid inputs, e.g. 1 and 1

**Comment by codewithkenzo** [comment]
I'd like to work on this.

After looking at the code and how other RPC frameworks handle this, I think re-throwing validation errors (rather than wrapping them) is the right approach:

1. gRPC, tRPC, JSON-RPC all distinguish validation errors (protocol-level) from tool execution errors (application-level)
2. MCP spec says `isError: true` is for tool errors, not validation failures
3. Early error throwing gives cleaner error messages

The fix would be in `McpServer.setToolRequestHandlers()` - for task-augmented requests, re-throw `InvalidParams` errors instead of wrapping them.

Let me know if maintainers prefer a different approach.

**Comment by sstklen** [comment]
Can confirm the repro — the underlying validation error gets completely masked by the `CreateTaskResultSchema` wrapping. You get an opaque schema error instead of the actual input validation failure, which makes debugging really painful.

@codewithkenzo's approach of re-throwing validation errors (rather than wrapping them as tool errors) aligns well with how gRPC and tRPC handle this distinction. Looking forward to the fix.
